zephyr/lib/cpu.c

```diff
@@ -153,17 +130,13 @@ int cpu_enable_core(int id)
 		z_init_cpu(id);
 #endif
 
-	atomic_clear(&start_flag);
 	atomic_clear(&ready_flag);
 
-	arch_start_cpu(id, z_interrupt_stacks[id], CONFIG_ISR_STACK_SIZE,
-		       secondary_init, &start_flag);
+	k_smp_cpu_start(id, secondary_init, NULL);
```

Comment:
Do you test it ? It has problems.   please check above code at line 128 to 130, we skip z_init_cpu(id) for some reason (check the code comments). But now k_smp_cpu_start ->start_cpu->z_init_cpu, so we call z_init_cpu again. This will break our context saving.

Reply:
@tmleman  please share your idea, thanks!

Reply:
Hm; z_init_cpu() should be idempotent when called again on a restarting core, all it's supposed to do it set up the cpu struct.  Nothing I can see would require skipping it, though maybe the newer obj_core stuff (which is the only non-trivial code there) might have an interaction, but that's brand new and I doubt SOF has turned it on to notice?

I'd hate to reject a patch like this based on "for some reason" on one platform.  If intel_adsp can't share the same resume path we should figure out why and fix it.

Reply:
Hmm, this is tricky. @andyross @tmleman this is documented in comments on L128-130, IMR_CONTEXT_SAVE is enabled and this is not the first boot, the CPU state is restored by soc code and calling z_init_cpu() would overwrite the restored stack. See Zephyr commit 3df442a982731ca4d8f2c1ad9508c1f157b789dc . Cc @ceolin as well.

Reply:
Heh. FWIW, I knew this synchronous resume trickery you guys were playing was going to be a problem.  I'll let you guys sort it out, but if the requirement at the platform layer is really as broad as "the entire state of the idle thread must be preserved across power transitions", I can guarantee you it'll be a perenial issue.   There's just no way to do that in a clean way given how many things need to happen on core shutdown/bringup in the abstract.  This isn't the way any other OS works either.

Can we **please** work out a way for the device to resume in a more reasonable way?  If you absolutely have to, we could work out a way to suspend your code in a thread context and then context switch directly into the idle thread to do the dirty work, then hook the callback here for hardware details and unsuspend your thread in the normal way.

Reply:
I guess at this point... just don't call `k_smp_cpu_start()` (or old `z_smp_cpu_start()`) when resuming, and only use that to power up the core for the very first time after boot. For resuming CPUs, just use `arch_start_cpu()`. Since the SoC layer has custom code when resuming, there is no need to involve the kernel.